File: kotlin/friends-tree/src/test/kotlin/com/pret/friend/tree/repository/UserFriendsRepositoryTest.kt
package com.pret.friend.tree.repository

import com.pret.friend.tree.data.DataComponent
import com.pret.friend.tree.abstractions.User
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class UserFriendsRepositoryTest {

    private val dataComponent: DataComponent = mockk()

    private val underTest = UserFriendsRepository(dataComponent)

    @Test
    fun willGetUserById() {
        every { dataComponent.users() } returns setOf(User(2, "Julian"))

        val result = underTest.getUserById(2)

        assertThat(result.id).isEqualTo(2)
        assertThat(result.name).isEqualTo("Julian")

        verify { dataComponent.users() }
    }

    @Test
    fun willUserFriends() {
        every { dataComponent.friendships() } returns setOf(Pair(2, 4))
        every { dataComponent.users() } returns setOf(User(4, "Joe"))

        val result = underTest.getFriendsById(2)

        assertThat(result).isEqualTo(listOf(User(4, "Joe")))

        verify { dataComponent.users() }
        verify { dataComponent.friendships() }
    }
}

File: nodejs/src/repository/userFriendsRepository.test.ts
import userFriendsRepository from "./userFriendsRepository";

describe("userFriendsRepository unit tests", () => {
  test("should get User By Id", () => {
    const user = userFriendsRepository.getUserById("1");
    expect(user).toEqual({ id: "1", name: "Alicia Keys" });
  });

  test("should get Friends By Id", () => {
    const friends = userFriendsRepository.getFriendsById("2");

    expect(friends.map((f) => f.id).sort()).toEqual(["3", "4", "7", "9"]);
  });
});
